server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessor.java

```diff
@@ -0,0 +1,294 @@
+            .filter(isNotLocalServer(context))
+            .toList();
+
+        if (!remoteAddress.isEmpty()) {
+            StringBuilder logBuffer = new StringBuilder(128);
+            try {
+                context.sendMail(sender, remoteAddress, message);
+                logBuffer.append("Mail for ")
+                    .append(recipient).append(" forwarded to ")
+                    .append(remoteAddress);
+            } catch (MessagingException ex) {
+                logBuffer.append("Error forwarding mail to ")
+                    .append(remoteAddress)
+                    .append("attempting local delivery");
+            }
+            context.log(LogLevel.INFO, logBuffer.toString());
+
+        }
+    }
+
+    private String getDefaultDomain(DomainList domainList) throws MessagingException {
+        try {
+            return domainList.getDefaultDomain();
+        } catch (DomainListException e) {
+            throw new MessagingException("Unable to access DomainList", e);
+        }
+    }
+
+    /**
+     * Handle the given mappings to map the original recipient to the right one
+     *
+     * @param mappings
+     *            a collection of mappings for the given recipient
+     * @param sender
+     *            the sender of the mail
+     * @param recipient
+     *            the original recipient of the email
+     * @param message
+     *            the mail message
+     * @return a collection of mapped recpient addresses
+     *
+     * @throws MessagingException
+     */
+    public Collection<MailAddress> handleMappings(Mappings mappings, MailAddress sender, MailAddress recipient, MimeMessage message,
+                                                  MailetContext context, DomainList domainList) throws MessagingException {
+        ImmutableList<Mapping> addressMappingWithoutDomains = getAddressWithDefaultDomain(mappings, domainList);
+
+        ImmutableList<Mapping> newAddressMappings = convertToNewMappings(mappings, addressMappingWithoutDomains);
+
+        ImmutableList<MailAddress> mailAddresses = buildMailAddressFromMappingAddress(newAddressMappings);
+
+        forwardToRemoteAddress(sender, recipient, message, mailAddresses, context);
+
+        return getLocalAddresses(mailAddresses, context);
+    }
+
+    public ImmutableList<MailAddress> rewriteRecipients(Mail mail, Function<MailAddress, Collection<MailAddress>> processMail) {
+        return FluentIterable
+            .from(mail.getRecipients())
+            .transformAndConcat(processMail)
+            .toList();
+    }
+
+    public void logAndAddError(Builder<MailAddress> errors, MailAddress recipient, Exception e, MailetContext context) {
+        context.log(LogLevel.INFO, "Error while process mail.", e);
+        errors.add(recipient);
+    }
+
+    public void replaceReturnPath(Mail mail) throws MessagingException {
+        mail.getMessage().setHeader(RFC2822Headers.RETURN_PATH, getReturnPath(mail.getSender()));
```

Comment:
It's done on separated commit. The return path should be edit by LocalDelivery and RemoteDelivery but so it means that RRTMailet must be set up with LocalDelivery and RemoteDelivery.